blockchain/blockdao/blockdao.go

```diff
@@ -231,16 +239,78 @@ func (dao *blockDAO) Footer(h hash.Hash256) (*block.Footer, error) {
 	return dao.footer(h)
 }
 
+func (dao *blockDAO) GetActionByActionHash(h hash.Hash256) (action.SealedEnvelope, error) {
+	if dao.indexer == nil {
+		return action.SealedEnvelope{}, blockindex.ErrActionIndexNA
+	}
+	actIndex, err := dao.indexer.GetActionIndex(h[:])
+	if err != nil {
+		return action.SealedEnvelope{}, err
+	}
+	bh, err := dao.getBlockHash(actIndex.BlockHeight())
+	if err != nil {
+		return action.SealedEnvelope{}, err
+	}
+	blk, err := dao.body(bh)
+	if err != nil {
+		return action.SealedEnvelope{}, err
+	}
+	for _, act := range blk.Actions {
+		if act.Hash() == h {
+			return act, nil
+		}
+	}
+	return action.SealedEnvelope{}, errors.Errorf("block %d does not have action %x", actIndex.BlockHeight(), h)
+}
+
+func (dao *blockDAO) GetReceiptByActionHash(h hash.Hash256) (*action.Receipt, error) {
+	if dao.indexer == nil {
+		return nil, blockindex.ErrActionIndexNA
+	}
+	actIndex, err := dao.indexer.GetActionIndex(h[:])
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get receipt index for action %x", h)
+	}
+	receipts, err := dao.getReceipts(actIndex.BlockHeight())
+	if err != nil {
+		return nil, err
+	}
+	for _, r := range receipts {
+		if r.ActionHash == h {
+			return r, nil
+		}
+	}
+	return nil, errors.Errorf("receipt of action %x isn't found", h)
+}
+
 func (dao *blockDAO) GetReceipts(blkHeight uint64) ([]*action.Receipt, error) {
 	return dao.getReceipts(blkHeight)
 }
 
 func (dao *blockDAO) PutBlock(blk *block.Block) error {
-	return dao.putBlock(blk)
+	if err := dao.putBlock(blk); err != nil {
+		return err
+	}
+	// index the block if there's indexer
+	if dao.indexer == nil {
+		return nil
+	}
+	if err := dao.indexer.PutBlock(blk); err != nil {
+		return err
+	}
+	return dao.indexer.Commit()
```

Comment:
We should unify the behavior of blockdao and indexer: PutBlock + Commit or PutBlock

Reply:
fixed in latest push